src/box/alter.cc

```diff
@@ -4031,28 +4031,13 @@ on_replace_dd_cluster(struct trigger *trigger, void *event)
 		on_rollback->data = replica;
 		txn_stmt_on_rollback(stmt, on_rollback);
 	} else {
-		/*
-		 * Don't allow deletion of the record for this instance
-		 * from _cluster.
-		 */
 		assert(old_tuple != NULL);
 		uint32_t replica_id;
 		if (tuple_field_u32(old_tuple, BOX_CLUSTER_FIELD_ID, &replica_id) != 0)
 			return -1;
 		if (replica_check_id(replica_id) != 0)
 			return -1;
-		/*
-		 * It's okay to update the instance id (drop, then insert) while
-		 * it is joining to a cluster as long as the id is set by the
-		 * time bootstrap is complete, which is checked in box_cfg()
-		 * anyway.
-		 *
-		 * For example, the replica could be deleted from the _cluster
-		 * space on the master manually before rebootstrap, in which
-		 * case it will replay this operation during the final join
-		 * stage.
-		 */
-		if (!replicaset.is_joining && replica_id == instance_id) {
+		if (replica_id == instance_id) {
```

Comment:
> replication: drop is_joining flag
>
> Its single usecase was to allow instance_id change during join.
> This was allowed assuming that at rebootstrap an instance will
> fetch not the latest master's state but its latest snapshot file.
> And then the changes from xlogs.
>
> Those xlogs could contain a deletion of this replica from _cluster
> if the master tried to clear the space from dead replicas.
>
> But join/reboostrap since long time ago don't work this way. The
> replica fetches the latest master's state right from memory, not
> from the latest snapshot file. Hence it won't see any deletions.
> Not even just from _cluster.
>
> That means this check can become stricter and is_joining flag is
> not needed anymore.

What does 'a long time ago' mean? The old relay (master) feeds snapshot+WAL data, correct? The old relay works in an unexpected way with the new replica, correct? I would like to know what is the version, there the data starts to send from the memory, and how exactly things is broken if the replay is older than this version. Maybe it also requires explicit 'breaking change' mark. Maybe it requires a note in the upgrade manual (at least [here](https://github.com/tarantool/tarantool/wiki/Upgrade-instructions)).

Reply:
In this case "a long time ago" means 4 years. Commit: https://github.com/tarantool/tarantool/commit/6332aca655ae7f95d391bdc0109e79915f6e6ad0, ticket: https://github.com/tarantool/tarantool/issues/1271.

> The old relay (master) feeds snapshot+WAL data, correct?

What do you mean as "old relay"?

> feeds snapshot+WAL data, correct?

Yes. If as snapshot you mean "latest master's state right from memory", not literally a snapshot file.

> The old relay works in an unexpected way with the new replica, correct?

I don't understand "old relay" part, but anyway - not correct. This commit doesn't do any functional changes. It deletes dead code.

> I would like to know what is the version, there the data starts to send from the memory

See links above.

> how exactly things is broken if the replay is older than this version

Things are not broken.

> Maybe it also requires explicit 'breaking change' mark. Maybe it requires a note in the upgrade manual

Neither a breaking nor a functional change. And doesn't need to be in the upgrade manual as this commit and those changes by the links above do not change any observable behaviour.

Reply:
> In this case "a long time ago" means 4 years. Commit: https://github.com/tarantool/tarantool/commit/6332aca655ae7f95d391bdc0109e79915f6e6ad0, ticket: https://github.com/tarantool/tarantool/issues/1271.

4 years is close to the time frame we want to support LTS releases in a future. 1.10 is supported for almost 5 years ATM. So it is not 'a long time ago', it is 'affecting supported versions'. Really same as 'a day ago' for us.

> What do you mean as "old relay"?

Relay from 1.10.15 for example (released less than 2 months ago). Well, not so old. Just relay without the linked commit.

I suppose that a user may upgrade from one supported version to another supported version and have working replication in process of upgrading.

OTOH, the patchset goes to 3.x only, right? I guess we can impose a restriction: 1.10 should be upgraded to 2.x first, then to 3.x. It seems fair enough.

Now, when the 'old relay' term is defined, whether it still looks as non-breaking/dead-code-elimination? Or it breaks upgrading from 1.10 to 3.x, but not upgrading from 2.3.1+ to 3.x? If so, can we add an explicit error on attempt to connect to 1.10/2.2/2.3.0 relay? An expicit error should be better than a bunch of subtle problems.

Reply:
I still don't understand what the problem is about, so I summarize what I was saying, don't have much to add:
* This commit (removing `is_joining` flag) is refactoring. It deletes dead code. The deleted branch was never hit after the ancient commit linked above.
* This entire patchset doesn't add any blockers for upgrading or replicating from/to < 3.0.0 versions, including 1.10. There might be some problems added before this patchset, but this one is not breaking in terms of protocols.

Unless I misunderstood you, you want me to mention here that my patchset breaks the protocols and/or to mention the old commit by the link provided above that it also breaks something. I can't do that, because 1) this patchset doesn't break the protocols, 2) that old commit by the link also doesn't break the protocols and is not related to my patchset.

Also I still consider "4 years" = long. "Long" is subjective and for me "4 years" is long.

Reply:
Oh, there is a problem indeed, if try to replicate from an instance < 2.3.1. Don't know if that is what you meant in (3), I still can't understand. But thanks for the persistence. The old commit about initial snapshot from memory is fine. Although I had to drop my commit about `is_joining` flag. Instead I managed to add a test which covers the case on all versions, including 3.0.0. I added it as a separate commit.

Reply:
It seems, we got each other in general and now there shouldn't be a problem with instances on tarantool < 2.3.1. I'm closing the discussion so.

Thank you for your attention!